### PR TITLE
Prevent old jetty-util-9.3.0 from showing up in IDE

### PR DESCRIPTION
**Pull request by joakime**
+ Since the `jetty-util-9.3.0.jar` reference is only for testing
  the webapp / war behavior of the WebAppClassloader, having
  that old jar as a `<dependency>` is not necessary.
  Moved it to a `maven-dependency-plugin:copy/artifactItem` instead.

Signed-off-by: Joakim Erdfelt <joakim.erdfelt@gmail.com>

### CODE REVIEW

**Comment by joakime** [comment]
Good point, I'll double check the OSGi manifest.
Ideally I don't want to see this as a full blown <dependency> as the dependency convergence in aggregate plugins (javadoc) and IDEs (eclipse and intellij) has a hard time with it that way.

**Comment by joakime** [comment]
Current `jetty-9.4.x` tree manifest (copied here for reference).

```
Manifest-Version: 1.0
Created-By: Apache Maven Bundle Plugin
Built-By: joakim
Build-Jdk: 11.0.3
Bnd-LastModified: 1560370328635
Bundle-ClassPath: .,WEB-INF/classes,WEB-INF/lib
Bundle-Copyright: Copyright (c) 2008-2019 Mort Bay Consulting Pty. Ltd.
Bundle-Description: Test Webapp for Servlet 3.1 Features
Bundle-DocURL: http://www.eclipse.org/jetty
Bundle-License: http://www.apache.org/licenses/LICENSE-2.0, http://www.e
 clipse.org/org/documents/epl-v10.php
Bundle-ManifestVersion: 2
Bundle-Name: Jetty Tests :: Webapps :: Spec Webapp
Bundle-RequiredExecutionEnvironment: JavaSE-1.8
Bundle-SymbolicName: org.eclipse.jetty.tests.test-spec-webapp
Bundle-Vendor: Eclipse Jetty Project
Bundle-Version: 9.4.20.SNAPSHOT
Export-Package: com.acme.test;version="9.4.20";uses:="com.acme.initializ
 er,javax.annotation,javax.annotation.security,javax.servlet,javax.servl
 et.annotation,javax.servlet.http,javax.sql"
Import-Package: javax.transaction;version="[1.1,1.3)",com.acme.initializ
 er;version="[9.4.20,10)",javax.annotation;version="[1.3.0,2)",javax.ann
 otation.security;version="[1.3.0,2)",javax.naming,javax.servlet;version
 ="[3.1.0,4)",javax.servlet.annotation;version="[3.1.0,4)",javax.servlet
 .http;version="[3.1.0,4)",javax.sql,org.eclipse.jetty.util;version="[9.
 3.0,10)",com.acme;version="9.4.20"
Jetty-ContextFilePath: /META-INF/plugin-context.xml
Require-Capability: osgi.ee;filter:="(&(osgi.ee=JavaSE)(version=1.8))"
Tool: Bnd-3.5.0.201709291849
Web-ContextPath: /
```

**Comment by joakime** [comment]
The manifest in branch `jetty-9.4.x-old-classloading-test`

```
Manifest-Version: 1.0
Created-By: Apache Maven Bundle Plugin
Built-By: joakim
Build-Jdk: 11.0.3
Bnd-LastModified: 1560370421943
Bundle-ClassPath: .,WEB-INF/classes,WEB-INF/lib
Bundle-Copyright: Copyright (c) 2008-2019 Mort Bay Consulting Pty. Ltd.
Bundle-Description: Test Webapp for Servlet 3.1 Features
Bundle-DocURL: http://www.eclipse.org/jetty
Bundle-License: http://www.apache.org/licenses/LICENSE-2.0, http://www.e
 clipse.org/org/documents/epl-v10.php
Bundle-ManifestVersion: 2
Bundle-Name: Jetty Tests :: Webapps :: Spec Webapp
Bundle-RequiredExecutionEnvironment: JavaSE-1.8
Bundle-SymbolicName: org.eclipse.jetty.tests.test-spec-webapp
Bundle-Vendor: Eclipse Jetty Project
Bundle-Version: 9.4.20.SNAPSHOT
Export-Package: com.acme.test;version="9.4.20";uses:="com.acme.initializ
 er,javax.annotation,javax.annotation.security,javax.servlet,javax.servl
 et.annotation,javax.servlet.http,javax.sql"
Import-Package: javax.transaction;version="[1.1,1.3)",com.acme.initializ
 er;version="[9.4.20,10)",javax.annotation;version="[1.3.0,2)",javax.ann
 otation.security;version="[1.3.0,2)",javax.naming,javax.servlet;version
 ="[3.1.0,4)",javax.servlet.annotation;version="[3.1.0,4)",javax.servlet
 .http;version="[3.1.0,4)",javax.sql,org.eclipse.jetty.util;version="[9.
 4.20,10)",com.acme;version="9.4.20"
Jetty-ContextFilePath: /META-INF/plugin-context.xml
Require-Capability: osgi.ee;filter:="(&(osgi.ee=JavaSE)(version=1.8))"
Tool: Bnd-3.5.0.201709291849
Web-ContextPath: /
```

It is using the 9.4.20 version in the manifest.
But why is this important?
Does OSGi really use the bundles on a webapp AND NOT the `WEB-INF/lib` ?

I'll address this anyway, but it still seems like a red-herring.

**Comment by joakime** [comment]
Update on commit b41f5b5cf4f1bf5d46274e7487e7237da86ec276 now produces the following corrected OSGi manifest ...

```
Manifest-Version: 1.0
Created-By: Apache Maven Bundle Plugin
Built-By: joakim
Build-Jdk: 11.0.3
Bnd-LastModified: 1560370635225
Bundle-ClassPath: .,WEB-INF/classes,WEB-INF/lib
Bundle-Copyright: Copyright (c) 2008-2019 Mort Bay Consulting Pty. Ltd.
Bundle-Description: Test Webapp for Servlet 3.1 Features
Bundle-DocURL: http://www.eclipse.org/jetty
Bundle-License: http://www.apache.org/licenses/LICENSE-2.0, http://www.e
 clipse.org/org/documents/epl-v10.php
Bundle-ManifestVersion: 2
Bundle-Name: Jetty Tests :: Webapps :: Spec Webapp
Bundle-RequiredExecutionEnvironment: JavaSE-1.8
Bundle-SymbolicName: org.eclipse.jetty.tests.test-spec-webapp
Bundle-Vendor: Eclipse Jetty Project
Bundle-Version: 9.4.20.SNAPSHOT
Export-Package: com.acme.test;version="9.4.20";uses:="com.acme.initializ
 er,javax.annotation,javax.annotation.security,javax.servlet,javax.servl
 et.annotation,javax.servlet.http,javax.sql"
Import-Package: javax.transaction;version="[1.1,1.3)",org.eclipse.jetty.
 util;version="[9.3.0,10)",com.acme.initializer;version="[9.4.20,10)",ja
 vax.annotation;version="[1.3.0,2)",javax.annotation.security;version="[
 1.3.0,2)",javax.naming,javax.servlet;version="[3.1.0,4)",javax.servlet.
 annotation;version="[3.1.0,4)",javax.servlet.http;version="[3.1.0,4)",j
 avax.sql,com.acme;version="9.4.20"
Jetty-ContextFilePath: /META-INF/plugin-context.xml
Require-Capability: osgi.ee;filter:="(&(osgi.ee=JavaSE)(version=1.8))"
Tool: Bnd-3.5.0.201709291849
Web-ContextPath: /
```

**Comment by gregw** [comment]
@joakime OSGi is meant to use dependencies as bundles.  We currently do fall back to WEB-INF/lib, but we are not sure that we should do so.
The manifest LGTM, but I'll defer to @janbartel  to confirm.

**Comment by janbartel** [comment]
The best thing to do is to take jetty-util-9.3 out of the dependencies altogether, and use some other means of getting it into WEB-INF/lib, eg with the dependency plugin.  

The manifest header "Bundle-Classpath" is telling OSGi that there are some classes that have been packed inside the bundle, that it should use *privately* to resolve against. Thus, we can just put the jetty-util-9.3 jar inside the webapp and not deploy it to the osgi container at all.

File: tests/test-webapps/test-servlet-spec/test-spec-webapp/pom.xml [hunk 1]
```diff
@@ -76,6 +76,7 @@
             <Import-Package>
               javax.transaction*;version="[1.1,1.3)",
               com.acme;version="${parsedVersion.majorVersion}.${parsedVersion.minorVersion}.${parsedVersion.incrementalVersion}",
+              org.eclipse.jetty.util;version="[9.3.0,10)",
```
Comment (janbartel) [inline tests/test-webapps/test-servlet-spec/test-spec-webapp/pom.xml]: don't add this entry for jetty-util, see my comment on the issue.

Comment (joakime) [reply tests/test-webapps/test-servlet-spec/test-spec-webapp/pom.xml]: I am using the dependency plugin, like your comment asked.
This entry brings the manifest for the `test-spec-webapp.war` back to way it was for the 9.4.19 release.
See my comments on the manifest history.  I have 3 of them.   
First manifest shown is the jetty-9.4.x branch (aka what 9.4.19 release looks like).
Second is the manifest as it was when this PR was first created, and Greg pointed out this mistake.
Third is the manifest after adding this `Import-Package` line. (bringing the manifest back to where Greg wanted it)

Comment (joakime) [reply tests/test-webapps/test-servlet-spec/test-spec-webapp/pom.xml]: resolved.

**Comment by janbartel** [comment]
@joakime I've had an offline discussion with @gregw: it's his preference that jetty-util remains a dependency in the pom.  Thus, we're going to fix this the "osgi way": 

* rollback to the pom.xml as it is in jetty-9.4.x
*  add the following 2 lines into the maven-bundle-plugin <instructions>:
```xml
            <Embed-Dependency>jetty-util</Embed-Dependency>
            <Embed-Directory>WEB-INF/lib</Embed-Directory>
```

Explanation: the above instructions tell osgi to include the jetty-util dependency jar inside the webapp bundle _and to resolve all requests for classes internally from it, without reference to the osgi container environment;_ it's like telling osgi that it has a private copy of the jar.

I've tested this against 9.4 and also 10 and this approach seems to be working and meet all the requirements.

**Comment by joakime** [comment]
That wont solve the dependency convergence issues on aggregate plugins (like javadoc).
We **must** remove the `<dependency>` entry for jetty-util 9.3.0 from the pom.xml.
That's required.
Rolling back to what's in jetty-9.4.x doesn't solve for that fundamental issue.

**Comment by gregw** [comment]
We ***must not*** remove the depency :)  at least not from some webapp somewhere.  We need to test that a webapp with an old dependency can be correctly deployed... but it can be an external test webapp.

**Pull request closed without merging**